### ISSUE TITLE
Unit conversion is not adhering to the specification

### ISSUE THREAD

**Issue by EvanMachusak** [FirelyTeam/firely-cql-sdk#144]
Our SDK performs unit equivalency for syntactic date units by normalizing them immediately to their UCUM counterparts.  

For example, [CqlDate line 208](https://github.com/FirelyTeam/firely-cql-sdk/blob/e69974a2ba15252c0d0fb17c82b9839df49ad07e/Cql/Cql.Primitives/CqlDate.cs#L208)

This behavior is contrary to the [Author's Guide §4.2.1 Quantities](https://cql.hl7.org/02-authorsguide.html#quantities).

In all cases, we should preserve the original unit and instead whenever it is required to compare units as equal or equivalent, we should invoke a not yet implemented UnitComparer that will compare UCUM and syntactic units per the specification.

Here's some CQL that could reproduce it:

```
define function "Is year precision"(quantity System.Quantity): quantity.unit = 'a'
define "Year equal to 'a'": "Is year precision"(1 year)
```

In our current implementation, `Year equal to 'a'` returns `true` where pursuant to the specification, it should return `false`:

> 1 year is not equal (=) to 1 'a' (defined in UCUM as 365.25 'd'), but it is equivalent (~) to 1 'a'.

**Comment by EvanMachusak** [comment]
Note:

```
define "Alpha": 1 day = 1 'd'
define "Beta": (1 day).unit = 'd'
```

`"Alpha"` is `true` in both the JavaScript and Java engines.
`"Beta"` is `false` in both the JavaScript and Java engines.

My proposed fix for this problem would result in `"Beta"` returning true, since we would do all unit comparisons with unit comparers, and `day` would be equal to `'d'`.

[Relevant zulip discussion thread here.](https://chat.fhir.org/#narrow/stream/179220-cql/topic/Unit.20equality.20clarifiaction)

**Comment by ewoutkramer** [comment]
That's right, in the current SDK Quantity I do not normalize, but I have instead introduced a system specifically for calendar durations (since they don't exist in UCUM).

https://github.com/FirelyTeam/firely-net-sdk/blob/develop/src/Hl7.Fhir.Base/ElementModel/Types/Quantity.cs

```
   public static Quantity ForCalendarDuration(decimal value, string calendarUnit)
        {
            return calendarUnit is not null
                ? new Quantity(value, calendarUnit, QuantityUnitSystem.CalendarDuration)
                : throw new ArgumentNullException(nameof(calendarUnit));
        }
```

Here, we introduced a QuantityUnitSystem.CalendarDuration - which also means that this CQL Quantity type *can* have a system that is not UCUM. 

This way, I don't have to normalize immediately, and can keep track of the "actual" unit.

**Comment by ewoutkramer** [comment]
Maybe we should use this system for it, which is specifically for calendar units: https://build.fhir.org/ig/HL7/FHIRPath/CodeSystem-calendar-units.html

**Comment by baseTwo** [comment]
## Still relevant — keeping open

Reviewed as part of a sweep over the oldest open issues (this one is from 2024). Rather than close it on age, here is what I checked and what is actually left.

### Assessment

Most of this is now fixed, but a concrete piece of the spec text quoted in the issue is still violated, so I would keep it open with the scope narrowed rather than close it. FIXED: units are no longer normalized to UCUM on construction - `CqlOperators.Quantity(1m, "year")` returns a CqlQuantity with unit "year" verbatim (generated code shows `context.Operators.Quantity(300m, "days")`), `Units.CqlUnitsToUCUM` (the dictionary the old CqlDate:208 went through) no longer exists, `CqlDate.Component` switches on "year"/"month"/"day" directly, and date arithmetic now distinguishes calendar from definite units (adding UCUM 'a'/'mo' signals CqlUcumYearArithmeticError/CqlUcumMonthArithmeticError). That came in via PR #1037 / issue #1024 (merged 2025-09-29). The "not yet implemented UnitComparer" the issue asks for also exists now: CqlQuantityCqlComparer canonicalizes through UCUM in both Compare and Equivalent, with a calendar<->UCUM mapping in UcumConversionExtensions. The issue's own repro no longer reproduces. STILL BROKEN: equality between the *variable-length* calendar durations and their UCUM counterparts. I measured `1 year = 1 'a'` -> true and `1 month = 1 'mo'` -> true, because CompareValues falls through to the canonicalizing branch which maps "year"->"a" before comparing. Spec 9.B (Equal) says "unit conversion between variable length calendar durations (i.e. years and months) and definite-time durations (i.e. days or below) results in null", and the Author's Guide sentence quoted verbatim in this issue says 1 year is not equal to 1 'a' (it is only equivalent). day/week/hour/minute/second/millisecond are correct (both = and ~ true, matching the spec table and the Java/JS engines as noted in the issue's first comment), and equivalence is correct across the board. A maintainer could reasonably instead close this and file a narrow successor titled something like "Equal (=) between year/month calendar durations and UCUM 'a'/'mo' should be null, not true" - the choice is presentational, but the defect itself is real and undecided (#1415 only settled equivalence, not this).

### Evidence

Unit preservation: Cql/Cql.Runtime/Operators/CqlOperators.TypeOperators.cs:388 returns `new CqlQuantity(value, unit)` unchanged; LibrarySets/dqm-content-qicore-2025/Extracted/CSharp/CMS56FHIRFuncStatHipReplacement-1.0.000.g.cs:1286 emits `Quantity(1m, "year")`. Calendar/UCUM arithmetic split: Cql/Cql.Abstractions/Primitives/CqlDate.cs:110-116. Comparer: Cql/Cql.Runtime/Comparers/CqlComparers.CqlQuantityCqlComparer.cs:37-93 (canonicalizing fallback at :49 and :86); mapping at Cql/Cql.Runtime/Conversion/UcumConversionExtensions.cs:128-142; equivalence proven by Cql/CoreTests/CqlComparersTests.cs:317-334. Fixed by PR #1037 (issue #1024), merged 2025-09-29, commit 282d5a64a. Measured on develop @ ca9166f39 with a throwaway CoreTests probe (since deleted; worktree left clean): `1 year vs 1 'a'` -> Compare=0, Equals=True, Equivalent=True; same for month/'mo'; day/'d' and week/'wk' -> True/True (correct); `new CqlQuantity(1m,"year").unit == "a"` -> False. Spec: spec/cql/condensed/09-b-cqlreference.md:1525 and :1537 (Equal -> null for year/month vs definite durations), :1593-1600 (Equivalent -> 1 year ~ 1 'a'), spec/cql/condensed/02-authorsguide.md:892 (the sentence quoted in the issue).

### Does it still reproduce?

Partly. The issue's own repro is fixed: `(1 year).unit` is now the string "year", so `"Is year precision"(1 year)` returns false as the spec requires - verified directly (`new CqlQuantity(1m, "year").unit == "a"` -> False), and generated library code confirms literals keep their calendar unit. What still deviates is equality: I measured `1 year = 1 'a'` and `1 month = 1 'mo'` returning true where spec 9.B requires null, via a temporary MSTest probe against CqlComparers on develop @ ca9166f39 (probe file deleted afterwards).

---

_Verified against `develop` at the time of writing; confidence: **medium**. If any of the above is wrong, the issue is a better candidate for closing than it looks._
